step-api/src/main/java/org/jenkinsci/plugins/workflow/steps/EnvironmentExpander.java

```diff
@@ -67,4 +69,32 @@ public static EnvironmentExpander merge(@CheckForNull EnvironmentExpander origin
         }
     }
 
+    /**
+     * Computes an effective environment in a given context.
+     * Used from {@code DefaultStepContext} and {@code EnvActionImpl}.
+     * The precedence order is:
+     * <ol>
+     * <li>{@code expander} (if any)
+     * <li>{@code customEnvironment}
+     * <li>{@code contextualEnvironment} (if any)
+     * </ol>
+     * @param customEnvironment {@link Run#getEnvironment(TaskListener)}, or {@code EnvironmentAction#getEnvironment}
+     * @param contextualEnvironment a possible override as per {@link BodyInvoker#withContext} (such as from {@link Computer#getEnvironment} called from {@code PlaceholderExecutable})
+     * @param expander a possible expander
+     * @return the effective environment
+     */
+    public static @Nonnull EnvVars getEffectiveEnvironment(@Nonnull EnvVars customEnvironment, @CheckForNull EnvVars contextualEnvironment, @CheckForNull EnvironmentExpander expander) throws IOException, InterruptedException {
+        EnvVars env;
+        if (contextualEnvironment != null) {
+            env = new EnvVars(contextualEnvironment);
+            env.putAll(customEnvironment);
+        } else {
+            env = new EnvVars(customEnvironment);
+        }
```

Comment:
Unfortunately those are not necessarily identical, because an `EnvVars` also holds a `platform` field which is left unset in the no-arg constructor. It is transferred in the copy constructor but only originally set in `masterEnvVars`, which in this case (assuming you are currently inside a `node` block, otherwise you do not care about the platform) would come via the `contextualEnvironment` passed from `PlaceholderExecutable` calling `Computer.getEnvironment` (which calls `getRemote`, which loads `masterEnvVars` in the slave JVM, meaning the slave’s environment variables, not the master’s).

If you forgot and started with the no-arg constructor, all of your tests would still pass, because `override` falls back to using `File.pathSeparatorChar` (“it's still more error prone that _[sic]_ I'd like” according to the comment), which from a test would be the same on a slave as on the master. But anything using `PATH+XXX` syntax might fail in the field.

Confused yet? I am, which is why I want to be done with refactoring this logic. :-)